File: apps/api/src/domain/schemas/patient.py
"""Patient Pydantic schemas for request/response validation."""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class PatientCreate(BaseModel):
    ahv_number: str | None = Field(None, pattern=r"^\d{3}\.\d{4}\.\d{4}\.\d{2}$", description="756.XXXX.XXXX.XX")
    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(..., min_length=1, max_length=100)
    date_of_birth: date
    gender: str = Field(..., pattern=r"^(male|female|other|unknown)$")
    blood_type: str | None = None
    phone: str | None = None
    email: str | None = None
    address_street: str | None = None
    address_zip: str | None = None
    address_city: str | None = None
    address_canton: str | None = Field(None, max_length=2)
    language: str = "de"


class PatientUpdate(BaseModel):
    first_name: str | None = None
    last_name: str | None = None
    phone: str | None = None
    email: str | None = None
    address_street: str | None = None
    address_zip: str | None = None
    address_city: str | None = None
    address_canton: str | None = None
    status: str | None = None


class PatientResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    ahv_number: str | None
    first_name: str
    last_name: str
    date_of_birth: date
    gender: str
    blood_type: str | None
    phone: str | None
    email: str | None
    address_street: str | None
    address_zip: str | None
    address_city: str | None
    address_canton: str | None
    language: str
    status: str
    created_at: datetime
    updated_at: datetime


class PaginatedPatients(BaseModel):
    items: list[PatientResponse]
    total: int
    page: int
    per_page: int

File: frontend/src/lib/validators/patient.ts
/**
 * Patient Zod schemas — mirrors backend Pydantic schemas.
 */
import { z } from "zod";

export const PatientCreateSchema = z.object({
  ahv_number: z.string().regex(/^756\.\d{4}\.\d{4}\.\d{2}$/, "Format: 756.XXXX.XXXX.XX").optional(),
  first_name: z.string().min(1, "Pflichtfeld").max(100),
  last_name: z.string().min(1, "Pflichtfeld").max(100),
  date_of_birth: z.string().min(1, "Pflichtfeld"),
  gender: z.enum(["male", "female", "other", "unknown"]),
  blood_type: z.string().optional(),
  phone: z.string().optional(),
  email: z.string().email("Ungültige E-Mail").optional().or(z.literal("")),
  address_street: z.string().optional(),
  address_zip: z.string().optional(),
  address_city: z.string().optional(),
  address_canton: z.string().max(2).optional(),
  language: z.string().default("de"),
});

export const PatientUpdateSchema = PatientCreateSchema.partial();

export type PatientCreateInput = z.infer<typeof PatientCreateSchema>;
export type PatientUpdateInput = z.infer<typeof PatientUpdateSchema>;
